src/interpolation/details/ArborX_InterpDetailsPolynomialBasis.hpp

```diff
@@ -0,0 +1,173 @@
+// And its recursive pattern:
+// - [1]                       at degree 0
+// - x*[1] and y*[1]           at degree 1
+// - x*[x] and y*[x, y]        at degree 2
+// - x*[xx] and y*[xx, xy, yy] at degree 3
+//
+// These examples can be represented in 2D as the following tables:
+//   Quadratic |  3D  ||        Cubic      |  2D
+// ------------+----- || ------------------+-----
+//    degree   | dim  ||       degree      | dim
+// ---+---+----+      || ---+---+----+-----+
+//  0 | 1 |  2 |      ||  0 | 1 |  2 |  3  |
+// ===o===o====o      || ===o===o====o=====o
+//  1 |        |      ||  1 |              |
+// ---+---+----+----- || ---+---+----+-----+-----
+//    | x |    |  x   ||    | x |    |     |  x
+//    |   | xx |      ||    |   | xx |     |
+//    +---+----+----- ||    |   |    | xxx |
+//    | y |    |  y   ||    +---+----+-----+-----
+//    |   | xy |      ||    | y |    |     |  y
+//    |   | yy |      ||    |   | xy |     |
+//    +---+----+----- ||    |   | yy |     |
+//    | z |    |  z   ||    |   |    | xxy |
+//    |   | xz |      ||    |   |    | xyy |
+//    |   | yz |      ||    |   |    | yyy |
+//    |   | zz |      ||
+
+// This function returns the lengths of the slices for a given maximum degree
+// and dimension.
+// Given a slice at degree n and dimension u, its length would be the sum of all
+// the slices' lengths of degree n-1 and of same or smaller dimension.
+//
+// Given the previous two examples, the two arrays would be:
+// Deg \ Dim | x | y | z  || Deg \ Dim | x | y
+// ----------+---+---+--- || ----------+---+---
+//      1    | 1 | 1 | 1  ||      1    | 1 | 1
+//      2    | 1 | 2 | 3  ||      2    | 1 | 2
+//                        ||      3    | 1 | 3
+template <std::size_t Dim, std::size_t Deg>
+KOKKOS_FUNCTION constexpr auto polynomialBasisSliceLengths()
+{
+  static_assert(
+      Deg != 0 && Dim != 0,
+      "Unable to compute slice lengths for a constant polynomial basis");
+
+  struct
+  {
+    std::size_t arr[Deg][Dim]{};
+  } result;
+  auto &arr = result.arr;
+
+  for (std::size_t dim = 0; dim < Dim; dim++)
+    arr[0][dim] = 1;
+
+  for (std::size_t deg = 0; deg < Deg; deg++)
+    arr[deg][0] = 1;
+
+  for (std::size_t deg = 1; deg < Deg; deg++)
+    for (std::size_t dim = 1; dim < Dim; dim++)
+      arr[deg][dim] = arr[deg - 1][dim] + arr[deg][dim - 1];
+
+  return result;
+}
+
+// This returns the size of the polynomial basis, which is the sum of all the
+// slices lengths and 1.
+// Both of the previous examples will result in a polynomial basis of size 10
+template <std::size_t Dim, std::size_t Deg>
+KOKKOS_FUNCTION constexpr std::size_t polynomialBasisSize()
+{
+  if constexpr (Deg != 0 && Dim != 0)
+  {
+    auto [arr] = polynomialBasisSliceLengths<Dim, Deg>();
+    std::size_t size = 1;
+
+    for (std::size_t deg = 0; deg < Deg; deg++)
+      for (std::size_t dim = 0; dim < Dim; dim++)
+        size += arr[deg][dim];
+
+    return size;
+  }
+  else
+  {
+    return 1;
+  }
```

Comment:
Let `f(dim, degree)` be the set of polynomials of (homogeneous) polynomials of degree `Degree` in `dim` variables. Then, 
```
g(dim, degree) = f(dim, degree) \ f(dim-1, degree)
```
is the set you are describing and
```
polynomialBasisSliceLengths<Dim, Degree>() = |g(dim, degree)|
```
Since the number of homogeneous polynomials is 
```
(Degree+Dim-1)!/(Degree!(Dim-1)!)
```
we get
```
(Degree+Dim-1)!/(Degree!(Dim-1)!) - (Degree+Dim-2)!/(Degree!(Dim-2)!) = (Degree+Dim-2)!/((Degree-1)!(Dim-1)!)
```
You can use this formula directly here instead of summing.

Reply:
If this formula is indeed correct (I haven't checked), we don't need the basis slice length function.